agent/php_execute.c

```diff
@@ -1966,20 +1967,40 @@ static void nr_php_observer_attempt_call_cufa_handler(NR_EXECUTE_PROTO) {
    * cause additional performance overhead, this should be considered a last
    * resort.
    */
-  const zend_op* prev_opline = execute_data->prev_execute_data->opline - 1;
+
+  /* 
+   * We cannot safely access the opline of internal functions, and we only
+   * want to instrument cufa calls from user functions anyway
+   */
+  if (UNEXPECTED(NULL == execute_data->prev_execute_data->func)) {
+    nrl_verbosedebug(NRL_AGENT, "%s: cannot get previous function", __func__);
+    return;
+  }
+  if (!ZEND_USER_CODE(execute_data->prev_execute_data->func->type)) {
+    nrl_verbosedebug(NRL_AGENT, "%s: caller is php internal function", __func__);
+    return;
+  }
+
+  const zend_op* prev_opline = execute_data->prev_execute_data->opline;
+
+  /* 
+   * Extra safety check. Previously, we instrumented by overwritting ZEND_DO_FCALL.
+   * Within OAPI, for consistency's sake, we will ensure the same
+   */
+  if (ZEND_DO_FCALL != prev_opline->opcode) {
+    nrl_verbosedebug(NRL_AGENT, "%s: cannot get previous function name", __func__);
+    return;
+  }
+  prev_opline -= 1; // Checks previous opcode
```

Comment:
Maybe add a comment explaining why it is safe to check previous opcode. https://github.com/newrelic/newrelic-php-agent/pull/639 has a good explanation:
> ZEND_DO_FCALL will never be the first opcode in an op array -- minimally, there is always at least a ZEND_INIT_FCALL before it -- so it is safe to iterate backwards in the opcode like execute_data->prev_execute_data->opline - 1


```suggestion
  /* ZEND_DO_FCALL will never be the first opcode in an op array -- minimally,
   * there is always at least a ZEND_INIT_FCALL before it -- so it is safe to 
   * check what previous opcode is. */
  prev_opline -= 1; // Checks previous opcode
```

Reply:
The large block comment above has this explanation https://github.com/newrelic/newrelic-php-agent/pull/708/files#diff-beedad7bbfc9687348e557b4752fbf6f54e4407a6e3177e8cabaaec7e3ffd625R1944